File: 2015/15/go/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type Ingredient struct {
	name       string
	capacity   int
	durability int
	flavor     int
	texture    int
	calories   int
}

func parse(data string) Ingredient {
	binding := strings.ReplaceAll(data, ",", "")
	parts := strings.Fields(binding)

	name := strings.TrimSuffix(parts[0], ":")

	capacity, err := strconv.Atoi(parts[2])
	if err != nil {
		panic("Error parsing capacity")
	}

	durability, err := strconv.Atoi(parts[4])
	if err != nil {
		panic("Error parsing durability")
	}

	flavor, err := strconv.Atoi(parts[6])
	if err != nil {
		panic("Error parsing flavor")
	}

	texture, err := strconv.Atoi(parts[8])
	if err != nil {
		panic("Error parsing texture")
	}

	calories, err := strconv.Atoi(parts[10])
	if err != nil {
		panic("Error parsing calories")
	}

	return Ingredient{
		name:       name,
		capacity:   capacity,
		durability: durability,
		flavor:     flavor,
		texture:    texture,
		calories:   calories,
	}
}

func calculateScore(amounts []int, ingredients []Ingredient, target_calories *int) int {
	capacity, durability, flavor, texture, calories := 0, 0, 0, 0, 0

	for i, amt := range amounts {
		capacity += amt * ingredients[i].capacity
		durability += amt * ingredients[i].durability
		flavor += amt * ingredients[i].flavor
		texture += amt * ingredients[i].texture
		calories += amt * ingredients[i].calories
	}

	if target_calories != nil && calories != *target_calories {
		return 0
	}
	if capacity < 0 {
		capacity = 0
	}
	if durability < 0 {
		durability = 0
	}
	if flavor < 0 {
		flavor = 0
	}
	if texture < 0 {
		texture = 0
	}

	return capacity * durability * flavor * texture
}

func findBestCombination(index int, remainingTeaspoons int, currentAmounts []int, ingredients []Ingredient, targetCalories *int, maxScore *int, bestAmounts *map[string]int) {
	if index == len(ingredients)-1 {
		currentAmounts[index] = remainingTeaspoons
		score := calculateScore(currentAmounts, ingredients, targetCalories)
		if score > *maxScore {
			*maxScore = score
			newMap := make(map[string]int)
			for i, ing := range ingredients {
				newMap[ing.name] = currentAmounts[i]
			}
			*bestAmounts = newMap
		}
		return
	}

	for teaspoons := 0; teaspoons <= remainingTeaspoons; teaspoons++ {
		currentAmounts[index] = teaspoons
		findBestCombination(index+1, remainingTeaspoons-teaspoons, currentAmounts, ingredients, targetCalories, maxScore, bestAmounts)
	}
}

func calculateMaxScore(ingredients []Ingredient, nSpoons int, targetCalories *int) (int, map[string]int) {
	maxScore := 0
	bestAmounts := make(map[string]int)
	currentAmounts := make([]int, len(ingredients))
	findBestCombination(0, nSpoons, currentAmounts, ingredients, targetCalories, &maxScore, &bestAmounts)
	return maxScore, bestAmounts
}

func main() {
	data, err := os.ReadFile("../input.txt")
	if err != nil {
		panic("Error reading input file")
	}
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")

	var ingredients []Ingredient
	for _, line := range lines {
		if strings.TrimSpace(line) == "" {
			continue
		}
		ing := parse(line)
		ingredients = append(ingredients, ing)
	}

	sep := strings.Repeat("=", 20)

	maxScore1, bestAmounts1 := calculateMaxScore(ingredients, 100, nil)
	fmt.Printf("%s Part 1 %s\nmax_score: %d\nbest_amounts: %v\n", sep, sep, maxScore1, bestAmounts1)

	calories := 500
	maxScore2, bestAmounts2 := calculateMaxScore(ingredients, 100, &calories)
	fmt.Printf("%s Part 2 %s\nmax_score: %d\nbest_amounts: %v\n", sep, sep, maxScore2, bestAmounts2)
}

File: 2015/15/rust/src/main.rs
use std::{collections::HashMap, fs};

struct Ingredients {
    name:String,
    capacity:i32,
    durability:i32,
    flavor:i32,
    texture:i32,
    calories:i32,
}

fn parse(data: &str) -> Ingredients {
    let binding = data.replace(",","");
    let data: Vec<&str> = binding.split_whitespace().collect();
    Ingredients {
        name: data[0].trim_end_matches(':').to_string(),
        capacity: data[2].parse::<i32>().expect("Error parsing capacity"),
        durability: data[4].parse::<i32>().expect("Error parsing durability"),
        flavor: data[6].parse::<i32>().expect("Error parsing flavor"),
        texture: data[8].parse::<i32>().expect("Error parsing texture"),
        calories: data[10].parse::<i32>().expect("Error parsing calories"),
    }
}

fn calculate_score(amounts: &[i32], ingredients: &[Ingredients], target_calories: Option<i32>) -> i32 {
    let num_ingredients = amounts.len();
    let mut capacity = 0;
    let mut durability = 0;
    let mut flavor = 0;
    let mut texture = 0;
    let mut calories = 0;

    for i in 0..num_ingredients {
        capacity += amounts[i] * ingredients[i].capacity;
        durability += amounts[i] * ingredients[i].durability;
        flavor += amounts[i] * ingredients[i].flavor;
        texture += amounts[i] * ingredients[i].texture;
        calories += amounts[i] * ingredients[i].calories;
    }

    if let Some(target) = target_calories {
        if calories != target {
            return 0;
        }
    }
    let properties = [capacity, durability, flavor, texture];
    properties.iter().map(|&p| p.max(0)).product()
}

fn find_best_combination(
    index: usize,
    remaining_teaspoons: i32,
    current_amounts: &mut Vec<i32>,
    ingredients: &[Ingredients],
    target_calories: Option<i32>,
    max_score: &mut i32,
    best_amounts: &mut HashMap<String, i32>,
) {
    if index == ingredients.len() -1 {
        current_amounts[index] = remaining_teaspoons;
        let score = calculate_score(current_amounts, ingredients, target_calories);

        if score > *max_score {
            *max_score = score;
            *best_amounts = ingredients
            .iter()
            .enumerate()
            .map(|(i, ing)| (ing.name.clone(), current_amounts[i]))
            .collect();
        }
        return;
    }
    for teaspoons in 0..=remaining_teaspoons {
        current_amounts[index] = teaspoons;
        find_best_combination(
            index + 1, 
            remaining_teaspoons - teaspoons, current_amounts, 
            ingredients, 
            target_calories, 
            max_score, 
            best_amounts
        );
    }
}

fn calculate_max_score(
    ingredients: &[Ingredients],
    n_spoons: i32,
    target_calories: Option<i32>,
) -> (i32, HashMap<String, i32>) {
    
    let mut max_score = 0;
    let mut best_amounts = HashMap::new();
    let mut current_amounts = vec![0; ingredients.len()];

    find_best_combination(0,
         n_spoons,
         &mut current_amounts, 
         ingredients, 
         target_calories, 
         &mut max_score,
         &mut best_amounts
        );
        (max_score, best_amounts)
}

fn main() {
    let data: Vec<Ingredients> = fs::read_to_string("../input.txt")
    .expect("Error reading input file")
    .trim()
    .lines()
    .map(|l| parse(l))
    .collect();

    let sep = "=".repeat(20);
    let (max_score_1, best_amounts_1) = calculate_max_score(&data, 100, None);
    println!("{} Part 1 {}\nmax_score: {}\nbest_amounts: {:?}", sep,sep, max_score_1, best_amounts_1);

    let (max_score_2, best_amounts_2) = calculate_max_score(&data, 100, Some(500));
    println!("{} Part 2 {}\nmax_score: {}\nbest_amounts: {:?}", sep,sep, max_score_2, best_amounts_2);
}
